identity-account-storage/src/storage/stronghold.rs

```diff
@@ -270,6 +280,98 @@ impl Storage for Stronghold {
       .map_err(Into::into)
   }
 
+  async fn encrypt_data(
+    &self,
+    did: &IotaDID,
+    data: Vec<u8>,
+    associated_data: Vec<u8>,
+    encryption_options: &EncryptionOptions,
+    private_key: &KeyLocation,
+    public_key: PublicKey,
+  ) -> Result<EncryptedData> {
+    let client: Client = self.client(&ClientPath::from(did))?;
+    match private_key.key_type {
+      KeyType::Ed25519 => Err(Error::InvalidPrivateKey(
+        "ED25519 keys are not suported for encrypting/decrypting data".to_owned(),
+      )),
+      KeyType::X25519 => {
+        let public_key: [u8; X25519::PUBLIC_KEY_LENGTH] = public_key
+          .as_ref()
+          .try_into()
+          .map_err(|_| Error::InvalidPublicKey(format!("expected type: [u8, {}]", X25519::PUBLIC_KEY_LENGTH)))?;
+        match encryption_options.cek_algorithm() {
+          CEKAlgorithm::ECDH_ES { agreement } => {
+            let shared_key: Location = diffie_hellman(&client, private_key, public_key).await?;
+            let concat_kdf: Location = concat_kdf(
+              &client,
+              &encryption_options.encryption_algorithm(),
+              "ECDH-ES".to_owned(),
+              agreement,
+              shared_key.clone(),
+            )
+            .await?;
+            let encrypted_data: Result<EncryptedData> = aead_encrypt(
+              &client,
+              &encryption_options.encryption_algorithm(),
+              concat_kdf.clone(),
+              data,
+              associated_data,
+            )
+            .await;
+            revoke_data(&client, shared_key, true).await?;
+            revoke_data(&client, concat_kdf, true).await?;
+            encrypted_data
+          }
+        }
+      }
+    }
+  }
+
+  async fn decrypt_data(
+    &self,
+    did: &IotaDID,
+    data: EncryptedData,
+    encryption_options: &EncryptionOptions,
+    private_key: &KeyLocation,
+    public_key: PublicKey,
+  ) -> Result<Vec<u8>> {
+    let client: Client = self.client(&ClientPath::from(did))?;
+    match private_key.key_type {
+      KeyType::Ed25519 => Err(Error::InvalidPrivateKey(
+        "ED25519 keys are not suported for encrypting/decrypting data".to_owned(),
+      )),
+      KeyType::X25519 => {
+        let public_key: [u8; X25519::PUBLIC_KEY_LENGTH] = public_key
+          .as_ref()
+          .try_into()
+          .map_err(|_| Error::InvalidPublicKey(format!("expected type: [u8, {}]", X25519::PUBLIC_KEY_LENGTH)))?;
+        match encryption_options.cek_algorithm() {
+          CEKAlgorithm::ECDH_ES { agreement } => {
+            let shared_key: Location = diffie_hellman(&client, private_key, public_key).await?;
+            let concat_kdf: Location = concat_kdf(
```

Comment:
@PhilippGackstatter are there chained procedures in Stronghold yet that we can use to avoid temporary locations?

Reply:
Looks like stronghold cannot do that, after all. I thought I remembered it did that once, but the procedures API has gone through more iterations since then, and this seems to have been dropped, or I misremembered completely. We will need to do the revocation ourselves, unfortunately.